src/Avalonia.Controls/Primitives/PopupPositioning/IPopupPositioner.cs

```diff
@@ -422,6 +425,51 @@ public enum PopupGravity
         BottomRight = Bottom | Right,
     }
 
+    /// <summary>
+    /// Represents a method that provides custom positioning for a <see cref="Popup"/> control.
+    /// </summary>
+    /// <param name="popupSize">The <see cref="Size"/> of the <see cref="Popup"/> control.</param>
+    /// <param name="targetRect">The <see cref="Rect"/> of the <see cref="Popup.PlacementTarget"/>.</param>
+    /// <param name="offset">The <see cref="Point"/> computed from the <see cref="Popup.HorizontalOffset"/> and <see cref="Popup.VerticalOffset"/> property values.</param>
+    public delegate CustomPopupPlacement CustomPopupPlacementCallback(Size popupSize, Rect targetRect, Point offset);
+
+    /// <summary>
+    /// Defines custom placement parameters for a Popup control.
+    /// </summary>
+    public record struct CustomPopupPlacement
```

Comment:
I would also break this out into a new file.

Additionally, if it was me, I would make is a class so it could actually be extended if ever needed. This is too complex of a type to lock-down to a struct in my opinion.

Reply:
In my mind struct is for primitives types (int, size, etc.) this is more complex. The reason to make it a class is to not limit yourself in the future (keep the door open for future extension by developers). There also doesn't appear to be a performance reason for this to be a struct.

So my criteria would be
 1. Everything is a class by default in an object-oriented language 
 2. Is it a primitive type? then make it a struct
 3. Is it performance critical? then make it a struct.
 4. Is it a special-case?

It fails both of these struct tests and isn't a special case so my convention would be to make it a class. By default types should be a class in an object oriented language like C# -- not a struct. I do think Avalonia went a little too far with struct usage elsewhere.